### ISSUE TITLE
Filename collisions in sort files

### ISSUE THREAD

**Issue by jodavies** [vermaseren/form#112]
Now that the sort files can be put in a different directory to the scratch files, (TempDir != TempSortDir), it is not safe to assume that sort files can be created with the same base filename.

For eg, if I start tform on a script which creates sort files with
#: TempDir /foo1

and another with
#: TempDir /foo2
#: TempSortDir /foo1

the second instance of tform will overwrite the existing xformxxx.sFN of the first, causing it to crash when it comes to merge the patches.

I suppose one way to avoid this is to always use -M, but this is not the default behaviour. Otherwise, threads must check each and every time they create a sort file.

Thanks,
Josh.

**Comment by vsht** [comment]
I'm kind of wondering how unique the sort files become when the -M option is used? Is there still a chance that when
running multiple tform processes and setting `FORMTMP` to `/tmp` a crash due to overwriting may occur?

I came up with my own workaround by using sth like
```
tfmTformTmpDir="$(mktemp -d)";
cleanup="yes"
tform -t "$tfmTformTmpDir" ...;
rm -rf $tfmTformTmpDir
```
which creates unique subfolders for each tform process. However, if there is a reliable built-in mechanism available,
I'd rather use that than reinventing the wheel.

**Comment by jodavies** [comment]
`-M` means the filename will use the first 5 digits of the `pid`. It is unlikely that there will be a collision, but I think it is not guaranteed.

**Comment by vsht** [comment]
I see, thanks. Then I'll try using that and see if it suits me. On the cluster this is never an issue, since each tform
process gets a unique environment, but on my laptop (Ryzen 4750U with 8C/16T) I have to take some precautions by hand.

**Comment by vermaseren** [comment]
In the -M option the temporary files have a name that contains the PID of the running (t)form.
This PID is supposed to be unique, UNLESS you have old files that were not removed
with an old PID in them of a job that ran some time ago. But that would mean that you loose
those old files. Normally that is no disaster, but in principle old .str files can be used in a
Load statement to recover their contents.

In the scheme without the PID, Form checks whether a .str file with the selected name exists
already. If so it raises the number in the name and tries again. This is also not failsafe, because
when you start programs very quickly after each other the .str file of the first job may not exist
yet when the second looks for that name. Also, sometimes when you delete the storage, the
file is temporarily removed and recreated.

**Comment by jodavies** [comment]
Indeed, on clusters usually there are no issues. On your laptop you would have to have two tform processes running whose `pid` differs by precisely 100000.

This is due to `startup.c`;s `int num = ((int)GetPID())%100000;`

**Comment by vermaseren** [comment]
This %100000 was put in the days that PID’s never went beyong 215 or 2^16. 
Nowadays it can indeed go beyond 10^5. At the same time the length of filenames
is far less critical these days. This limitation can be taken ot I would say. The only thing
is that the space reserved for these filenames should get a variable length, which may
need a few modifications in several places. Something for the next version I would say.

**Comment by jodavies** [comment]
This PR is vaguely related and proposes the same scheme for the names https://github.com/vermaseren/form/pull/284 .

**Comment by vsht** [comment]
> But that would mean that you loose
those old files. Normally that is no disaster, but in principle old .str files can be used in a
Load statement to recover their contents.

In fact, large leftover sort files that remain in the case of FORM crashing or aborting the evaluation
is something I was always wondering about. Of course, from the developer point of view those
can be quite valuable for debugging, but I guess that an ordinary user is never interested in them
except from the fact that they need to be removed by hand. 

Perhaps as a suggestion for some future version, it would be cool to have a command line argument
telling FORM to remove **all** temporary files before terminating.

**Comment by vermaseren** [comment]
In principle Form removes all temporaries, even when it crashes. But there are crashes
in which it does not get the chance to do so.
For removing those I have made an alias in the shell:
alias xxx rm xform* whatevertmpdirectory/xform*
This way the xxx shell command cleans up all those files.

**Comment by vsht** [comment]
I see, thanks.
 
Of course, if one runs FORM from a shell script and that script creates a unique TEMP directory
for FORM, the script will as well remove it after the FORM process ends (gracefully or not).

**Comment by tueda** [comment]
Technically, maybe `mkstemp` + `unlink` would work for automatically removing all temporaries (https://github.com/vermaseren/form/issues/14#issuecomment-278291226), but using this idea in the FORM source code can be highly nontrivial...

**Comment by jodavies** [comment]
This is mitigated by -M as the default behaviour in v5 (#591)
